### ISSUE TITLE
Not possible to restore margins of a layer-shell window after setting them to large enough negative values

### ISSUE THREAD

**Issue by Tamaranch** [labwc/labwc#499]
Labwc 0.5.3 and git-master on Arch Linux

Here is the code to reproduce the problem:

<p>
<details>
<summary>p.c</summary>

```c
#include <gtk/gtk.h>
#include <gtk-layer-shell/gtk-layer-shell.h>

#define WINDOW_SIZE 100

static gint offset = 0;

static gboolean
on_button_press (GtkWidget *widget,
                 GdkEvent  *event,
                 gpointer   user_data)
{
  gtk_layer_set_margin (user_data, GTK_LAYER_SHELL_EDGE_TOP, - WINDOW_SIZE + offset);

  return TRUE;
}

static gboolean
on_button_release (GtkWidget *widget,
                   GdkEvent  *event,
                   gpointer   user_data)
{
  gtk_layer_set_margin (user_data, GTK_LAYER_SHELL_EDGE_TOP, 0);

  return TRUE;
}

gint main (gint argc, gchar **argv)
{
  GtkWidget *window;
  GMainLoop *loop;
  GtkWidget *widget;

  if (argv[1] != NULL)
    offset = atoi (argv[1]);

  gtk_init (&argc, &argv);

  window = gtk_window_new (GTK_WINDOW_TOPLEVEL);
  widget = gtk_button_new_with_label ("test");
  gtk_container_add (GTK_CONTAINER (window), widget);
  g_signal_connect (widget, "button-press-event", G_CALLBACK (on_button_press), window);
  g_signal_connect (widget, "button-release-event", G_CALLBACK (on_button_release), window);

  gtk_layer_init_for_window (GTK_WINDOW (window));
  gtk_layer_set_anchor (GTK_WINDOW (window), GTK_LAYER_SHELL_EDGE_TOP, TRUE);
  gtk_widget_set_size_request (window, WINDOW_SIZE, WINDOW_SIZE);

  gtk_widget_show (widget);
  gtk_widget_show (window);

  loop = g_main_loop_new (NULL, FALSE);
  g_signal_connect_swapped (window, "destroy", G_CALLBACK (g_main_loop_quit), loop);
  g_main_loop_run (loop);
  g_main_loop_unref (loop);
}
```

</details>
</p>

It is built by
```sh
gcc p.c -o p $(pkg-config --cflags --libs gtk+-3.0 gtk-layer-shell-0)
```

Run by
```sh
./p [offset]
```
Pressing the button applies a negative top margin (by default of the window size), releasing it resets the margin to zero. 

Passing a positive offset as input works, but it doesn't work for an offset less than or equal to zero. It works in all cases on Wayfire.

For info my use case is hiding a panel by moving it off screen on X11, which I ported to Wayland this way: https://gitlab.xfce.org/xfce/xfce4-panel/-/merge_requests/96

**Comment by johanmalm** [comment]
Thanks. I can re-produce, but don’t know what’s causing it. Will investigate.

**Comment by Consolatis** [comment]
Unrelated to this issue:

> It seems very preferable or even necessary to start the compositor from a tty to make proper tests.

labwc (or more precise wlroots) has a X11 backend as well, so it should start up fine on an existing X11 session (possibly using another user / dbus session though).

> The two main compositors used for this work are Labwc and Wayfire.

There is also MIR which appears to have implemented `wlr-foreign-toplevel-management-unstable-v1` and `wlr-layer-shell-unstable-v1` [among others](https://github.com/MirServer/mir/tree/main/src/wayland/protocol).

**Comment by Tamaranch** [comment]
> labwc (or more precise wlroots) has a X11 backend as well, so it should start up fine on an existing X11 session

Yes it starts up fine, as do the other compositors I tested, but I noticed that their behavior was not the same as when run from a tty.

For example the test case given in #475 does not work in Wayfire when run from an X11 session, while it works fine when run from a tty. Same with Sway, although in this case the test with layer-shell doesn't work, but the difference in behavior without layer-shell also exists.

I seem to have come across other such differences, related to event handling, but I don't remember which ones. Anyway I concluded that to get the "real behavior" of a compositor, it was better to launch it from a tty.

> There is also MIR

Thanks! I haven't tried it yet, but I probably will at some point.

**Comment by Consolatis** [comment]
We did some debugging in wlroots and so far my theory is that it is an rendereing optimization in wlroots that is too broad: If the surface is not placed within some output (e.g. the exact use-case you desire) it won't trigger a commit while rendering and thus we will never even know that the margins have changed.

Your example works fine if you have two outputs and use the connecting edge to anchor the example to. Because when pressing the button the full layer is now showing up on the second output and thus the commit triggers when releasing the button and the surface moves back to its original place.

So there indeed seems to be an issue in wlroots when a layershell client uses a margin that is >= its height/width and there is no additional output in that direction. I would actually argue that negative margins should be forbidden completely because they may leak into neighboring outputs.

I am wondering though if the whole idea of using positioning to hide the panel is the best solution, why not just hide it instead?
e.g. something like this:

```c
gtk_widget_hide(GTK_WIDGET(actual_panel_window));
/* replace with an invisible hitbox */
gtk_widget_show(GTK_WIDGET(transparent_background_panel_without_exclusive_zone));

/* when detecting a cursor entering the replacement surface */
gtk_widget_hide(GTK_WIDGET(transparent_background_panel_without_exclusive_zone));
gtk_widget_show(GTK_WIDGET(actual_panel_window));
```
`transparent_background_panel_without_exclusive_zone` would then just act as a hitbox so you know when to restore the original panel on mouse over. Or similar: `gtk_widget_show`ing two new layershell surfaces, each only containing an arrow button on he bottom left and right which on click restore the original panel and hide themselves.

**Comment by Tamaranch** [comment]
> Your example works fine if you have two outputs and use the connecting edge to anchor the example to.

Interesting: I had not tested this configuration with Labwc, but I had tested it with Wayfire. For the latter, the window is not transferred to the other output, it disappears then reappears when the button is clicked, exactly as if there was no second output. I confirm the behavior you describe for Labwc though.

So which behavior is "the right one"? (is there one that is the right one?)

> I am wondering though if the whole idea of using positioning to hide the panel is the best solution, why not just hide it instead?

The first reason is: because it's currently done that way in the panel code on X11, and I want to change this code as little as possible when porting to Wayland. In particular, there is an animation when the panel is "hidden", by moving it progressively off screen.

Then, beyond the animation, it may be that it was done that way for a reason, rather than `gtk_widget_hide()` that first comes to mind. I wonder if hiding the panel, and thus all the plugins in it, doesn't affect their behavior under certain circumstances.

**Comment by Consolatis** [comment]
> So which behavior is "the right one"? (is there one that is the right one?)

Having the surface show up on a neighboring output seems a bit strange to me personally.
Not coming back when updating the margin is def. a bug though.

After some more research I found [this](https://gitlab.freedesktop.org/wayland/wayland-protocols/-/merge_requests/28#note_1341314) in the MR to turn the wlr-layershell protocol into an official wayland protocol. It mentions the [viewporter protocol](https://gitlab.freedesktop.org/wayland/wayland-protocols/-/blob/main/stable/viewporter/viewporter.xml) which could be an alternative solution.

**Comment by Tamaranch** [comment]
Nice find, thanks, I'll follow up on that. I might try using the viewporter protocol at some point, but negative margins are really the easiest and most practical thing for me. I'm glad to see I'm not the only one doing it by the way :)

**Comment by Consolatis** [comment]
This seems to work-around the bug by the way:

```c
static gboolean
on_button_release (GtkWidget *widget,
                   GdkEvent  *event,
                   gpointer   user_data)
{
  gtk_layer_set_margin (user_data, GTK_LAYER_SHELL_EDGE_TOP, 0);
  gtk_widget_hide(GTK_WIDGET(user_data));
  gtk_widget_show(GTK_WIDGET(user_data));

  return TRUE;
}
```

**Comment by Tamaranch** [comment]
I have to move the pointer out and in again to be able to click the button after that (no event sent otherwise).

**Comment by Consolatis** [comment]
Yeah, its not perfect by any stretch of the imagination. Its an pretty ugly work-around, we will try to get the underlying issue fixed in wlroots.

**Comment by Tamaranch** [comment]
It seems that this problem can and perhaps should be solved on the client side. See the issues listed above. I'll let you judge whether this should be closed.
